### PR TITLE
Perl6: Add many more smartmatch examples. Make the ternary operator clearer

### PR DESCRIPTION
**Pull request by samcv**
Add many more smartmatch examples. Make the ternary operator clearer and add a code example for it. Make the section for && and || have a working code example and show the output it gives.
If this is all good so far I will keep making changes.  Let me know what you think.

- [x] PR touches only one file (or a set of logically related files with similar changes made)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
- [x] YAML Frontmatter formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
  - [x] Seriously, look at it now. Watch for quotes and double-check field names.

### CODE REVIEW

File: perl6.html.markdown [hunk 1]
```diff
@@ -193,6 +193,15 @@ sub mutate($n is rw) {
   say "\$n is now $n !";
 }
 
+my $m = 42;
+mutate $m; # $n is now 43 !
```
Comment (vendethiel) [inline perl6.html.markdown]: I don't think this should be here.

File: perl6.html.markdown [hunk 2]
```diff
@@ -228,10 +237,15 @@ unless False {
 say "Quite truthy" if True;
 
 # - Ternary conditional, "?? !!" (like `x ? y : z` in some other languages)
-my $a = $condition ?? $value-if-true !! $value-if-false;
+# - returns $value-if-true if the condition is true and $value-if-false
+# - if it is false.
+#   my $result = $value condition ?? $value-if-true !! $value-if-false;
```
Comment (vendethiel) [inline perl6.html.markdown]: that line is unnecessary

File: perl6.html.markdown [hunk 3]
```diff
@@ -228,10 +237,15 @@ unless False {
 say "Quite truthy" if True;
 
 # - Ternary conditional, "?? !!" (like `x ? y : z` in some other languages)
-my $a = $condition ?? $value-if-true !! $value-if-false;
+# - returns $value-if-true if the condition is true and $value-if-false
+# - if it is false.
```
Comment (vendethiel) [inline perl6.html.markdown]: no need for those "-"

File: perl6.html.markdown [hunk 4]
```diff
@@ -336,16 +351,37 @@ if long-computation() -> $result {
...
-# - `~~` is smart matching
+# - Smart Match Operator: `~~`
+# Aliases the left hand side to $_ and then evaluates the right hand side.
+# Here are some common comparison semantics:
+
+# Type - Checks if left side "has type" (can check superclasses and roles)
+
+1 ~~ Int; # True
+
+# Smart-matching against a boolean always returns that boolean (and will warn).
```
Comment (vendethiel) [inline perl6.html.markdown]: Could you move this one at the end of the examples, seeing as it's a special case of sorts?

File: perl6.html.markdown (hunk 1 above)
Comment (samcv) [inline perl6.html.markdown]: I put it there to show how passing a variable works with 'is rw' and on the next few lines I show how it doesn't work on passing an integer.  If you think this part is unnecessary I can remove it

File: perl6.html.markdown (hunk 2 above)
Comment (samcv) [inline perl6.html.markdown]: Which one? 242?

**Comment by vendethiel** [comment]
That was really quick :). Is it good to merge?

**Comment by samcv** [comment]
Yes it should be good to merge :)
Also, thoughts on making the whole code valid?  It would be nice is the .pl file that is downloaded would be able to run.  Currently it doesn't because variables and sub's are redefined.  I fixed one section in 981162b

**Comment by vendethiel** [comment]
> Also, thoughts on making the whole code valid? It would be nice is the .pl file that is downloaded would be able to run. Currently it doesn't because variables and sub's are redefined. I fixed one section in 981162b

I don't care too much. Getting better names would be nice; tho.

**Comment by vendethiel** [comment]
👍

**Pull request merged**